### DISCUSSION TITLE
extending new languages for CPG

### DISCUSSION THREAD

**Discussion by anisyusof-sc** [Fraunhofer-AISEC/cpg#597] [Ideas]
hi there! is there any quick guide on how can we extend the library to support new languages in the CPG library?

my team has an intention to extend this CPG library to support PowerShell and would appreciate some guidance on this.

thank you in advance!

**Comment by oxisto** [comment]
Unfortunately, we do not have a tutorial on how to do this. In general you can have a look at the existing code. The core idea is the following:

The basis of each language is a `LanguageFrontend`, which is also the base java/kotlin class each language frontend should extend. The entry point so to speak is the `parse` function which takes an individual file as input and needs to return its CPG representation through a `TranslationUnitDeclaration`. 

# Parsing

Then you basically need to go through the AST (ideally with an existing parser framework) and translate the language-specific AST nodes into CPG nodes. Although not mandatory, I would suggest following the `Handler` pattern of the existing language frontends which basically translate groups of nodes, e.g. declarations to their CPG counterpart, e.g. `Declaration`. The handlers have a built-in switch map to execute handler functions based on the class of the raw AST node. You might need to switch manually on an AST node type property, if they all have the same class. This is for example used in the handlers of the `TypeScriptLanguageFrontend`.

https://github.com/Fraunhofer-AISEC/cpg/blob/0b599fe4b913d30f1f4e25148f5ab1cfc0a0aee9/cpg-library/src/main/java/de/fraunhofer/aisec/cpg/frontends/typescript/DeclarationHandler.kt#L37-L42

# Scoping

It is important to know that the CPG needs some kind of information about how scoping works in the language. The main interaction point with that is the `ScopeManager`, which is available to each language frontend. There are three important functions to know:
* `enterScope` can be called on a node to begin a new scope. This is mostly used on a `FunctionDeclaration` to specify that everything within the function is within its function scope or for a `CompoundStatement` to introduce the concept of variables local to the block. There is a special case however. In order to produce a global / file level scope, it is important to call `resetToGlobal` instead of `enterScope` on the main `TranslationUnitDeclaration`. This is because there is usually just *one* global scope.
* `leaveScope` needs to be called for each `enterScope` to close the scope - with exception of the global scope
* `addDeclaration` adds a declaration to the `DeclarationHolder` of the current scope. This may be un-initiative at first glance but is very powerful. For example if you create a new `ParamVariableDeclaration`, instead of figuring out where the function that this belongs to is located in the CPG, you just need to call `addDeclaration` using the new parameter declaration and the scope manager will automatically attach it to the function whose scope is currently active.

**Reply by anisyusof-sc** [reply]
thank you for the detailed quick start. we will have a look into this.
